src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+        };
+        Some((UpsertKey::from_value(value.as_ref(), &key_indices), value))
+    });
+    let (output_handle, output) = builder.new_output();
+
+    // An output that just reports progress of the snapshot consolidation process upstream to the
+    // persist source to ensure that backpressure is applied
+    let (_snapshot_handle, snapshot_stream) =
+        builder.new_output::<CapacityContainerBuilder<Vec<Infallible>>>();
+
+    let (mut health_output, health_stream) = builder.new_output();
+    let mut input = builder.new_input_for(
+        &input.inner,
+        Exchange::new(move |((key, _, _), _, _)| UpsertKey::hashed(key)),
+        &output_handle,
+    );
+
+    let mut persist_input = builder.new_disconnected_input(
+        &persist_input.inner,
+        Exchange::new(|((key, _), _, _)| UpsertKey::hashed(key)),
+    );
+
+    let upsert_shared_metrics = Arc::clone(&upsert_metrics.shared);
+
+    let shutdown_button = builder.build(move |caps| async move {
+        let [mut output_cap, snapshot_cap, health_cap]: [_; 3] = caps.try_into().unwrap();
+        let mut snapshot_cap = CapabilitySet::from_elem(snapshot_cap);
+
+        // The order key of the `UpsertState` is `Option<FromTime>`, which implements `Default`
+        // (as required for `consolidate_snapshot_chunk`), with slightly more efficient serialization
+        // than a default `Partitioned`.
+
+        let mut state = UpsertState::<_, Option<FromTime>>::new(
+            state_fn().await,
+            upsert_shared_metrics,
+            &upsert_metrics,
+            source_config.source_statistics,
+            upsert_config.shrink_upsert_unused_buffers_by_ratio,
+        );
+
+        // True while we're still reading the initial "snapshot" (a whole bunch
+        // of updates, all at the same initial timestamp) from our persist
+        // input or while we're reading the initial snapshot from the upstream
+        // source.
+        let mut hydrating = true;
+
+        // A re-usable buffer of changes, per key. This is an `IndexMap` because it has to be `drain`-able
+        // and have a consistent iteration order.
+        let mut commands_state: indexmap::IndexMap<_, upsert_types::UpsertValueAndSize<Option<FromTime>>> =
+            indexmap::IndexMap::new();
+        let mut multi_get_scratch = Vec::new();
+
+        // For our source input, both of these.
+        let mut stash = vec![];
+        let mut input_upper = Antichain::from_elem(Timestamp::minimum());
+
+
+        // For our persist/feedback input, both of these.
+        let mut persist_stash = vec![];
+        let mut persist_upper = Antichain::from_elem(Timestamp::minimum());
+
+        // A buffer for our output.
+        let mut output_updates = vec![];
+
+        let mut error_emitter = (&mut health_output, &health_cap);
+        let mut legacy_errors_to_correct = vec![];
+
+
+        loop {
+            tokio::select! {
+                Some(persist_event) = persist_input.next() => {
```

Comment:
We should use the `.ready()` and `.next_sync()` method below to avoid having to go through the now_or_never() machinery, in the interest of code clarity